contracts/bonding/BondingCheckpoints.sol

```diff
@@ -0,0 +1,393 @@
+        );
+
+        if (rewardRound < bond.lastClaimRound) {
+            // If the transcoder hasn't called reward() since the last time the delegator claimed earnings, there wil be
+            // no rewards to add to the delegator's stake so we just return the originally bonded amount.
+            return bond.bondedAmount;
+        }
+
+        (uint256 stakeWithRewards, ) = EarningsPoolLIP36.delegatorCumulativeStakeAndFees(
+            startPool,
+            endPool,
+            bond.bondedAmount,
+            0
+        );
+        return stakeWithRewards;
+    }
+
+    /**
+     * @notice Returns the last initialized earning pool for a transcoder at a given round.
+     * @dev Transcoders are just delegators with a self-delegation, so we find their last checkpoint before or at the
+     * provided _round and use its lastRewardRound value to grab the calculated earning pool. The only case where this
+     * returns a zero earning pool is if the transcoder had never called reward() before _round.
+     * @param _transcoder Address of the transcoder to look for
+     * @param _round Past round at which we want the valid earning pool from
+     * @return rewardRound Round in which the returned earning pool was calculated.
+     * @return pool EarningsPool.Data struct with the last initialized earning pool.
+     */
+    function getTranscoderLastRewardsEarningPool(address _transcoder, uint256 _round)
+        internal
+        view
+        returns (uint256 rewardRound, EarningsPool.Data memory pool)
+    {
+        BondingCheckpoint storage bond = getBondingCheckpointAt(_transcoder, _round);
+        rewardRound = bond.lastRewardRound;
+
+        // only fetch pool if there is a previous reward() call recorded
+        if (rewardRound > 0) {
+            pool = getTranscoderEarningPoolForRound(_transcoder, rewardRound);
+        }
+    }
+
+    /**
+     * @dev Proxy for {BondingManager-getTranscoderEarningsPoolForRound} that returns an EarningsPool.Data struct.
+     */
+    function getTranscoderEarningPoolForRound(address _transcoder, uint256 _round)
+        internal
+        view
+        returns (EarningsPool.Data memory pool)
+    {
+        (
+            pool.totalStake,
+            pool.transcoderRewardCut,
+            pool.transcoderFeeShare,
+            pool.cumulativeRewardFactor,
+            pool.cumulativeFeeFactor
+        ) = bondingManager().getTranscoderEarningsPoolForRound(_transcoder, _round);
+
+        if (pool.cumulativeRewardFactor == 0) {
+            revert MissingEarningsPool(_transcoder, _round);
+        }
+    }
+
+    /**
+     * @dev Helper to return more helpful custom errors in case of bad queries.
+     */
+    function checkedFindLowerBound(uint256[] storage array, uint256 value) internal view returns (uint256) {
+        if (array.length == 0) {
+            revert NoRecordedCheckpoints();
+        } else if (array[0] > value) {
+            revert PastLookup(value, array[0]);
```

Comment:
Should this be folded into SortedArrays since it seems that a bunch of checks are already implemented there?

Reply:
Fixed on https://github.com/livepeer/protocol/pull/614/commits/d5062fe2db7d5037988c2f1a5989886d7ce087d3